### PR TITLE
BUG: pd.concat produces frames with inconsistent order when concating the ones with categorical indices

### LINKED ISSUE #44099: BUG: `pd.concat` produces frames with inconsistent order when concating the ones with categorical indices
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas

df1 = pandas.DataFrame(
    {"col1": ["a_val", "b_val", "c_val"]},
    index=pandas.CategoricalIndex(["a", "b", "c"], categories=["a", "b", "c"]),
)
df2 = pandas.DataFrame(
    {"col1": ["b_val", "a_val", "c_val"]},
    # The same categories as in the first frame, but the values are in a different order
    index=pandas.CategoricalIndex(["b", "a", "c"], categories=["b", "a", "c"]),
)

print(pandas.concat([df1, df2]))

# Output:
#     col1
# a  a_val
# b  b_val
# c  c_val
# a  b_val <-- index doesn't math the original value
# b  a_val <-- index doesn't math the original value
# c  c_val
```

### Issue Description

When concatenating frames with categorical indices that have the same values but they are in a different order, the order of indices does not match the order of values from the original frame. 

In the reproducer above each "x" index value corresponds to the "x_val" value in the frame ("a" -> "a_val", "b" -> "b_val", ...), but in the concated frame this order is broken, the index values appears to be sorted, when the actual frame's values don't.

### Expected Behavior

An expected result would be either properly sorted result:
```python
    col1
a  a_val
b  b_val
c  c_val
a  a_val
b  b_val
c  c_val
```
or the one that is unsorted at all:
```python
    col1
a  a_val
b  b_val
c  c_val
b  b_val
a  a_val
c  c_val

```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-65-generic
Version          : #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : None
pytest           : 6.2.5
hypothesis       : None
sphinx           : 4.2.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : None
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fsspec           : 2021.08.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : 3.0.8
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : 2021.08.1
scipy            : None
sqlalchemy       : 1.4.23
tables           : None
tabulate         : None
xarray           : 0.19.0
xlrd             : 2.0.1
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by tyuyoshi**
- [ ] closes #44099 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/indexes/category.py [hunk 1]
```diff
@@ -571,7 +571,10 @@ def map(self, mapper):
     def _concat(self, to_concat: list[Index], name: Hashable) -> Index:
         # if calling index is category, don't check dtype of others
         try:
-            codes = np.concatenate([self._is_dtype_compat(c).codes for c in to_concat])
+            data = np.concatenate(
```
Comment (tyuyoshi) [inline pandas/core/indexes/category.py:574]: In the issue(https://github.com/pandas-dev/pandas/issues/44099), it uses `pd.concat` method.

In the concat method, it call `indexes.append` (https://github.com/pandas-dev/pandas/blob/1bd193eebad6897989bbf06ae62adcffefdf97cd/pandas/core/reshape/concat.py#L683).

In the `append` method, id call `_concat`.

`append`
https://github.com/pandas-dev/pandas/blob/1bd193eebad6897989bbf06ae62adcffefdf97cd/pandas/core/indexes/base.py#L5271
`_concat`
https://github.com/pandas-dev/pandas/blob/1bd193eebad6897989bbf06ae62adcffefdf97cd/pandas/core/indexes/category.py#L571

And i found the bug in `_concat` method.

File: pandas/tests/indexes/categorical/test_category.py [hunk 2]
```diff
@@ -286,6 +286,17 @@ def test_map_str(self):
         # See test_map.py
         pass
 
+    def test_append(self):
```
Comment (tyuyoshi) [inline pandas/tests/indexes/categorical/test_category.py:289]: The bug occurs in C`ategoricalIndex.append` method.
So i write the test here.

File: doc/source/whatsnew/v1.5.0.rst [hunk 3]
```diff
@@ -267,6 +267,7 @@ Categorical
 ^^^^^^^^^^^
 - Bug in :meth:`Categorical.view` not accepting integer dtypes (:issue:`25464`)
 - Bug in :meth:`CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
+- Bug in :meth:`CategoricalIndex._concat` when concatenating categorical indexes that have the same categories returning indexes in improper order (:issue:`44099`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:270]: can you refer to `pd.concat` as that's the most user visible bug here

**Review by jreback — CHANGES_REQUESTED** [review]
can you add the example from the original OP as well and put a test in pandas/tests/reshape/test_concat.py

File: doc/source/whatsnew/v1.5.0.rst (hunk 3 above)
Comment (tyuyoshi) [reply doc/source/whatsnew/v1.5.0.rst]: Hi, thank you for review!

I added description in this commit(https://github.com/pandas-dev/pandas/pull/46019/commits/aacc75c6807b882f6142007f8802d92f80815a74)

**Comment by tyuyoshi** [comment]
> can you add the example from the original OP as well and put a test in pandas/tests/reshape/test_concat.py

Thanks! I added the test case in this commit(https://github.com/pandas-dev/pandas/pull/46019/commits/a66a60fe6284b7380303d514cacfac14f29693d8)

File: pandas/tests/reshape/concat/test_concat.py [hunk 4]
```diff
@@ -502,6 +503,26 @@ def test_concat_duplicate_indices_raise(self):
...
+            {"col1": ["b_val", "a_val", "c_val"]},
+            index=CategoricalIndex(["b", "a", "c"], categories=["b", "a", "c"]),
+        )
+        expected = DataFrame(
+            {"col1": ["a_val", "b_val", "c_val", "b_val", "a_val", "c_val"]},
+            index=CategoricalIndex(
+                ["a", "b", "c", "b", "a", "c"], categories=["a", "b", "c"]
+            ),
+        )
+        tm.assert_frame_equal(concat([df1, df2]), expected)
```
Comment (jreback) [inline pandas/tests/reshape/concat/test_concat.py:524]: use 
```result = concat([df1, df2])```

File: pandas/tests/indexes/categorical/test_category.py [hunk 5]
```diff
@@ -286,6 +286,17 @@ def test_map_str(self):
...
+        # GH 44099
+        # concat indexes which have the same categories
+
+        ci1 = CategoricalIndex(["a", "b", "c"], categories=["a", "b", "c"])
+        ci2 = CategoricalIndex(["b", "a", "c"], categories=["b", "a", "c"])
+        expected = CategoricalIndex(
+            ["a", "b", "c", "b", "a", "c"], categories=["a", "b", "c"]
+        )
+        tm.assert_index_equal(ci1.append(ci2), expected)
```
Comment (jreback) [inline pandas/tests/indexes/categorical/test_category.py:298]: use ```result=ci1.append(ci2)```

File: doc/source/whatsnew/v1.5.0.rst [hunk 6]
```diff
@@ -267,6 +267,7 @@ Categorical
 ^^^^^^^^^^^
 - Bug in :meth:`Categorical.view` not accepting integer dtypes (:issue:`25464`)
 - Bug in :meth:`CategoricalIndex.union` when the index's categories are integer-dtype and the index contains ``NaN`` values incorrectly raising instead of casting to ``float64`` (:issue:`45362`)
+- Bug in :meth:`CategoricalIndex._concat` when concatenating categorical indexes that have the same categories returning indexes in improper order; this method can be called in :meth:`concat` (:issue:`44099`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:270]: can you update the note here; this should be about the user facing case in the OP

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/concat/test_concat.py (hunk 4 above)
Comment (tyuyoshi) [reply pandas/tests/reshape/concat/test_concat.py]: Thanks! I changed at this commit https://github.com/pandas-dev/pandas/pull/46019/commits/91d215568867985d1014f680d60d41f61f33af1e

File: pandas/tests/indexes/categorical/test_category.py (hunk 5 above)
Comment (tyuyoshi) [reply pandas/tests/indexes/categorical/test_category.py]: I also changed at this commit https://github.com/pandas-dev/pandas/commit/91d215568867985d1014f680d60d41f61f33af1e

File: doc/source/whatsnew/v1.5.0.rst (hunk 6 above)
Comment (tyuyoshi) [reply doc/source/whatsnew/v1.5.0.rst]: Thanks! I added description line at `reshaping` section that refer to `concat` func in this commit https://github.com/pandas-dev/pandas/pull/46019/commits/3425b854ca9f3c9370ad1cfb36f03d3ba9b58000

Is this the correct fix that you pointed out?

Comment (tyuyoshi) [reply doc/source/whatsnew/v1.5.0.rst]: @jreback Hi! Thanks for all your reviews.

I would like to request a review from you again.
I finally change [doc/source/whatsnew/v1.5.0.rst](https://github.com/pandas-dev/pandas/pull/46019/files/aacc75c6807b882f6142007f8802d92f80815a74#diff-c84bc2c71f43444548da8b2eb7b2f34efb50dac71eeddabbfcdd8ae7d7badc4b) in this commit([0e62a36](https://github.com/pandas-dev/pandas/pull/46019/commits/0e62a362b761f5069980f29aef0e40a5f024225b))

File: pandas/tests/reshape/concat/test_concat.py [hunk 7]
```diff
@@ -502,6 +503,27 @@ def test_concat_duplicate_indices_raise(self):
...
+            {"col1": ["b_val", "a_val", "c_val"]},
+            index=CategoricalIndex(["b", "a", "c"], categories=["b", "a", "c"]),
+        )
+        expected = DataFrame(
+            {"col1": ["a_val", "b_val", "c_val", "b_val", "a_val", "c_val"]},
+            index=CategoricalIndex(
+                ["a", "b", "c", "b", "a", "c"], categories=["a", "b", "c"]
+            ),
+        )
+        result = concat([df1, df2])
```
Comment (jreback) [inline pandas/tests/reshape/concat/test_concat.py:524]: can you make an asv for this type of concatanation and then show how it performs previously. i am worried that the list conversion of the codes is very expensive.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to be stale. If interested in continuing, please merge in the main branch, address the review, and we can reopen.

**Pull request closed without merging**